pkg/bgpv1/agent/controller.go

```diff
@@ -187,10 +200,22 @@ func (c *Controller) Run(ctx context.Context) {
 			return
 		case <-c.Sig.Sig:
 			l.Info("Cilium BGP Control Plane Controller woken for reconciliation")
+			if retrying {
+				if err := bo.Wait(ctx); err != nil {
+					l.Info("Context cancelled during backoff")
+					return
+				}
+			}
 			if err := c.Reconcile(ctx); err != nil {
 				l.WithError(err).Error("Encountered error during reconciliation")
+				c.Sig.Event(struct{}{}) // signal again to retry
+				retrying = true
```

Comment:
What are the cases where retrying reconciliation will resolve the returned error. We may need to categorize errors which can be retried and which can not. I am not sure if we have practical examples of errors which will be resolved when retried.

Reply:
Yeah, this was what I thought during the development, TBH. It is easy to introduce the error classification. We can introduce an error type like [resiliency.Retryable](https://github.com/cilium/cilium/blob/d721b8be7df8372acc3832498638f8c5a699f0f5/pkg/resiliency/error.go#L12), and we can handle it with `errors.Is` and friends.

What we need to decide here is a default behavior (retry by default vs no retry by default). I thought retrying by default was a "safer" option, but yes, practically, we don't have much situation in our current reconcilers that can be recovered by retry. 

The typical case is the error that comes from a temporal IO error (like network or storage), but at the time of reconciliation, we already fetched the k8s resources into the memory, so it doesn't matter; we don't touch the storage; so it doesn't matter. BGP Daemon does retry for any error by itself asynchronously, so it doesn't matter. etc...

In that sense, I don't have any objection to going with no retry by default and introducing `Retryable` error handling later as needed.

Reply:
I agree on the distinction between retryable and non-retryable errors. And yeah, ATM probably non-retryable could be the default, as majority of our reconcile work is configuring the process-local BGP instance. That would change, if we later decided to go with a BGP process running outside of the agent - I guess in that case most of the errors should be retried.

One error that potentially could be retryable even now is creation of a new gobgp server - that can temporarily fail e.g. if an old instance of the server is still running / was just killed etc. and the bind port is not yet free.

Reply:
As we agreed in the offline conversation, we'll reduce the scope of this PR and discuss separately about retry. Let me close this discussion. Currently, we decided to go with no retry by default.